build/templates/grpc_session_options.rst.mako

```diff
@@ -0,0 +1,105 @@
+<%
+    import build.helper as helper
+
+    config = template_parameters['metadata'].config
+    module_name = config['module_name']
+    driver_name = config['driver_name']
+    enums = config['enums']
+%>\
+${helper.get_rst_header_snippet('gRPC Support', '=')}
+
+Support for using ${driver_name} over gRPC
+
+.. py:currentmodule:: ${module_name}
+
+
+Enums
+=====
+
+
+SessionInitializationBehavior
+-----------------------------
+
+.. py:class:: SessionInitializationBehavior
+
+    .. py:attribute:: SessionInitializationBehavior.AUTO
+
+
+
+        Attach to an existing session with the specified name if it exists, otherwise intialize a new session.
+
+        .. note:: If this initializes a new session on the NI gRPC Device Server, then when the Python session goes out of scope,
+            it will automatically close the server session. If this attaches to an existing session on the NI gRPC Device Server,
+            then when the Python session goes out of scope, it will detach from the server session and leave it open.
```

Comment:
I've seen us talk about "Session going out of scope". I don't think this is right. `__exit()__` only gets called when the object was created as a context manager. Customers may and do sometimes use these Session objects by invoking the constructor "normally".

Consider:

```
class ContextManager():
    def __init__(self):
        print('init method called')

    def __enter__(self):
        print('enter method called')
        return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        print('exit method called')

    def __del__(self):
        print('del method called')


print("Test as context manager")

with ContextManager() as manager:
    print('with statement block')

print("\nTest as regular variable")

if True:
    c = ContextManager();

print("The end")
```

outputs:

```
Test as context manager
init method called
enter method called
with statement block
exit method called

Test as regular variable
init method called
The end
del method called
del method called
```

I think we may want to ensure that the appropriate thing happens on `__del()__` if it `__exit()__` never got called.
We also need the documentation to cover both cases.
We still (in documentation and examples) encourage customers to use as context manager.

We may need to run this through a Python guru.

Reply:
I looked at the .html documentation for Session and mimicked the verbiage about context managers. So I think it's much more technically precise now. I don't think you need to say much about what happens if the session is neither deleted nor exited. To me, that's _caveat emptor_.